algorithms/src/snark/marlin/ahp/ahp.rs

```diff
@@ -258,67 +352,165 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
-        let g_b_at_gamma = evals.get_lc_eval(&g_b, gamma)?;
-        let selector_b = largest_non_zero_domain.evaluate_selector_polynomial(non_zero_b_domain, gamma);
-        let lhs_b =
-            Self::construct_lhs("b", alpha, beta, gamma, v_H_at_alpha * v_H_at_beta, g_b_at_gamma, *sum_b, selector_b);
-        matrix_sumcheck += (r_b, &lhs_b);
-
-        let g_c = LinearCombination::new("g_c", [(F::one(), "g_c")]);
-        let g_c_at_gamma = evals.get_lc_eval(&g_c, gamma)?;
-        let selector_c = largest_non_zero_domain.evaluate_selector_polynomial(non_zero_c_domain, gamma);
-        let lhs_c =
-            Self::construct_lhs("c", alpha, beta, gamma, v_H_at_alpha * v_H_at_beta, g_c_at_gamma, *sum_c, selector_c);
-        matrix_sumcheck += (r_c, &lhs_c);
+        for (i, (&id, circuit_state)) in state.circuit_specific_states.iter().enumerate() {
+            let v_H_i_at_alpha_time = start_timer!(|| format!("v_H_i_at_alpha {id}"));
+            let v_H_i_at_alpha = circuit_state.constraint_domain.evaluate_vanishing_polynomial(alpha);
+            end_timer!(v_H_i_at_alpha_time);
+
+            let v_H_i_at_beta_time = start_timer!(|| format!("v_H_i_at_beta {id}"));
+            let v_H_i_at_beta = circuit_state.constraint_domain.evaluate_vanishing_polynomial(beta);
+            end_timer!(v_H_i_at_beta_time);
+            let v_H_i_alpha_beta = v_H_i_at_alpha * v_H_i_at_beta;
+
+            let selector_a = Self::get_selector_evaluation(
+                &mut cached_selectors,
+                &largest_non_zero_domain,
+                &circuit_state.non_zero_a_domain,
+                gamma,
+            );
+            let selector_b = Self::get_selector_evaluation(
+                &mut cached_selectors,
+                &largest_non_zero_domain,
+                &circuit_state.non_zero_b_domain,
+                gamma,
+            );
+            let selector_c = Self::get_selector_evaluation(
+                &mut cached_selectors,
+                &largest_non_zero_domain,
+                &circuit_state.non_zero_c_domain,
+                gamma,
+            );
+
+            let g_a_label = witness_label(id, "g_a", 0);
+            let g_a = LinearCombination::new(g_a_label.clone(), [(F::one(), g_a_label)]);
+            let g_b_label = witness_label(id, "g_b", 0);
+            let g_b = LinearCombination::new(g_b_label.clone(), [(F::one(), g_b_label)]);
+            let g_c_label = witness_label(id, "g_c", 0);
+            let g_c = LinearCombination::new(g_c_label.clone(), [(F::one(), g_c_label)]);
+
+            Self::add_g_m_term(
+                &mut matrix_sumcheck,
+                id,
+                "a",
+                alpha,
+                beta,
+                gamma,
+                evals,
+                &g_a,
+                v_H_i_alpha_beta,
+                r_a[i],
+                sums[&id].sum_a,
+                selector_a,
+            )?;
```

Comment:
bleh this is getting formatted really badly... 

Is there a way for us to combine some of the arguments (e.g.: pack `alpha`, `beta`, `gamma` into a 3-tuple called verifier challenges?)

Reply:
Fixed in b17840352. Used a struct so we still still name them clearly